File: backend/app/engine/document_rules.py
from typing import List, Dict, Any


def generate_document_requirements(profile: Any, approvals: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
    approval_names = {a["name"] for a in approvals}
    docs = []

    # 1. Restaurant Specific Documents
    if "FSSAI Registration / Licence" in approval_names:
        docs.append({
            "name": "Food Safety Management System (FSMS) Plan",
            "category": "Food Safety",
            "file_type": "PDF",
            "file_size": "2.1 MB",
            "status": "Verified",
            "required": True,
            "upload_date": "10 Mar 2024",
            "expiry_date": "Annual Audit",
            "renewal_cycle": "Annual"
        })
        docs.append({
            "name": "Food Handler Medical Fitness Certificates",
            "category": "Food Safety",
            "file_type": "PDF",
            "file_size": "1.4 MB",
            "status": "Verified",
            "required": True,
            "upload_date": "15 Feb 2024",
            "expiry_date": "15 Feb 2025",
            "renewal_cycle": "Annual"
        })

    if "Health Trade Licence" in approval_names:
        docs.append({
            "name": "Water Potability & Chemical Analysis Report",
            "category": "Health & Hygiene",
            "file_type": "PDF",
            "file_size": "1.8 MB",
            "status": "Verified",
            "required": True,
            "upload_date": "05 Jan 2024",
            "expiry_date": "05 Jan 2025",
            "renewal_cycle": "6 Months"
        })

    if "Eating House Licence" in approval_names:
        docs.append({
            "name": "Premises Seating Blueprint & Police Verification",
            "category": "Police Licensing",
            "file_type": "PDF",
            "file_size": "3.5 MB",
            "status": "Pending",
            "required": True,
            "upload_date": "12 Apr 2024",
            "expiry_date": "Perpetual",
            "renewal_cycle": "Perpetual"
        })

    if "Fire Safety NOC (Kitchen / Dining)" in approval_names:
        docs.append({
            "name": "Kitchen Exhaust Hood & Fire Extinguisher Clearance Certificate",
            "category": "Safety",
            "file_type": "PDF",
            "file_size": "2.8 MB",
            "status": "Pending",
            "required": True,
            "upload_date": "18 Apr 2024",
            "expiry_date": "18 Apr 2025",
            "renewal_cycle": "Annual"
        })

    # 2. Textile Specific Documents
    if "State Pollution Control Board Consent (Air & Water / ETP)" in approval_names:
        docs.append({
            "name": "ETP Consent to Operate (CTO) & Effluent Quality Report",
            "category": "Environmental",
            "file_type": "PDF",
            "file_size": "4.2 MB",
            "status": "Verified",
            "required": True,
            "upload_date": "10 Jan 2024",
            "expiry_date": "31 Dec 2026",
            "renewal_cycle": "5 Years"
        })

    if "Textile Committee / Ministry of Textiles Compliance" in approval_names:
        docs.append({
            "name": "Textile Committee Statutory Registration Certificate",
            "category": "Statutory Industry Compliance",
            "file_type": "PDF",
            "file_size": "1.2 MB",
            "status": "Verified",
            "required": True,
            "upload_date": "20 Jan 2024",
            "expiry_date": "Perpetual",
            "renewal_cycle": "Perpetual"
        })

    # 3. Jewellery Specific Documents
    if "BIS Hallmarking Registration" in approval_names:
        docs.append({
            "name": "BIS Hallmarking Scheme Registration Certificate",
            "category": "Certification",
            "file_type": "PDF",
            "file_size": "1.6 MB",
            "status": "Verified",
            "required": True,
            "upload_date": "15 Feb 2024",
            "expiry_date": "Perpetual",
            "renewal_cycle": "5 Years"
        })
        docs.append({
            "name": "Assaying & Hallmarking Centre Linkage Agreement",
            "category": "Certification",
            "file_type": "PDF",
            "file_size": "2.2 MB",
            "status": "Pending",
            "required": True,
            "upload_date": "20 Apr 2024",
            "expiry_date": "20 Apr 2026",
            "renewal_cycle": "Annual"
        })

    if "PMLA FIU-IND Reporting Registration" in approval_names:
        docs.append({
            "name": "FIU-IND Reporting Entity Registration Proof & AML Policy Document",
            "category": "Financial Compliance",
            "file_type": "PDF",
            "file_size": "1.9 MB",
            "status": "Verified",
            "required": True,
            "upload_date": "05 Mar 2024",
            "expiry_date": "Perpetual",
            "renewal_cycle": "Perpetual"
        })
        docs.append({
            "name": "Strong Room / Vault Armoury & Physical Security Audit Certificate",
            "category": "Physical Security",
            "file_type": "PDF",
            "file_size": "3.1 MB",
            "status": "Verified",
            "required": True,
            "upload_date": "12 Mar 2024",
            "expiry_date": "12 Mar 2025",
            "renewal_cycle": "Annual"
        })

    # 4. Retail Specific Documents
    if "Legal Metrology (Weights & Measures) Packaged Commodities Registration" in approval_names:
        docs.append({
            "name": "Legal Metrology Verification Certificate & Stamping Seal",
            "category": "Statutory Standards",
            "file_type": "PDF",
            "file_size": "1.3 MB",
            "status": "Verified",
            "required": True,
            "upload_date": "22 Jan 2024",
            "expiry_date": "22 Jan 2025",
            "renewal_cycle": "Annual"
        })

    if "Signage Board Display Permission" in approval_names:
        docs.append({
            "name": "Municipal Signage Board Dimension Drawing & Tax Receipt",
            "category": "Municipal Permitting",
            "file_type": "PDF",
            "file_size": "2.7 MB",
            "status": "Verified",
            "required": True,
            "upload_date": "14 Feb 2024",
            "expiry_date": "31 Mar 2026",
            "renewal_cycle": "Annual"
        })

    # 5. Factory / Manufacturing Documents
    if "Factory License" in approval_names:
        docs.append({
            "name": "Factory Licence Certificate",
            "category": "Statutory License",
            "file_type": "PDF",
            "file_size": "2.4 MB",
            "status": "Verified",
            "required": True,
            "upload_date": "14 Apr 2024",
            "expiry_date": "31 Dec 2026",
            "renewal_cycle": "Annual"
        })
        docs.append({
            "name": "Approved Factory Machinery Layout Plan",
            "category": "Statutory License",
            "file_type": "PDF",
            "file_size": "5.1 MB",
            "status": "Verified",
            "required": True,
            "upload_date": "10 Feb 2024",
            "expiry_date": "Perpetual",
            "renewal_cycle": "Perpetual"
        })

    if "Pollution Control NOC (CTO)" in approval_names:
        docs.append({
            "name": "Consent to Operate (CTO) Renewal Application",
            "category": "Environmental",
            "file_type": "PDF",
            "file_size": "4.8 MB",
            "status": "Pending",
            "required": True,
            "upload_date": "02 May 2024",
            "expiry_date": "Under Review",
            "renewal_cycle": "5 Years"
        })
        docs.append({
            "name": "Environmental Impact Assessment & Hazard Plan",
            "category": "Environmental",
            "file_type": "PDF",
            "file_size": "3.8 MB",
            "status": "Required",
            "required": True,
            "upload_date": None,
            "expiry_date": "Required for Audit",
            "renewal_cycle": "Annual"
        })

    if "Boiler / Pressure Vessel Licence" in approval_names:
        docs.append({
            "name": "Boiler Inspection & Hydraulic Test Certificate",
            "category": "Industrial Safety",
            "file_type": "PDF",
            "file_size": "2.5 MB",
            "status": "Verified",
            "required": True,
            "upload_date": "18 Mar 2024",
            "expiry_date": "18 Mar 2025",
            "renewal_cycle": "Annual"
        })

    # 6. Common General Documents
    if "Shop & Establishment Registration" in approval_names or "Shops & Establishments Registration" in approval_names:
        docs.append({
            "name": "Shops & Establishments Registration (Form C)",
            "category": "Business Registration",
            "file_type": "PDF",
            "file_size": "1.5 MB",
            "status": "Verified",
            "required": True,
            "upload_date": "15 Jan 2024",
            "expiry_date": "Perpetual",
            "renewal_cycle": "Perpetual"
        })

    if "GST Registration & Compliance" in approval_names or "GST Registration" in approval_names:
        docs.append({
            "name": "GST Registration Certificate (REG-06)",
            "category": "Taxation",
            "file_type": "PDF",
            "file_size": "1.1 MB",
            "status": "Verified",
            "required": True,
            "upload_date": "10 Jan 2024",
            "expiry_date": "Perpetual",
            "renewal_cycle": "Perpetual"
        })

    if "Municipal Trade Licence" in approval_names or "Local Trade Licence" in approval_names:
        docs.append({
            "name": "Municipal Trade Licence Endorsement",
            "category": "Business Operation",
            "file_type": "PDF",
            "file_size": "1.7 MB",
            "status": "Verified",
            "required": True,
            "upload_date": "05 Feb 2024",
            "expiry_date": "31 Mar 2026",
            "renewal_cycle": "Annual"
        })

    if "Fire Safety Certificate (Industrial)" in approval_names or "Fire Safety Certificate" in approval_names or "Fire Safety NOC" in approval_names:
        docs.append({
            "name": "Fire Safety Extinguisher & Sprinkler Test Report",
            "category": "Safety",
            "file_type": "PDF",
            "file_size": "3.2 MB",
            "status": "Pending",
            "required": True,
            "upload_date": "15 May 2024",
            "expiry_date": "15 May 2025",
            "renewal_cycle": "Annual"
        })

    if "MSME Udyam Registration" in approval_names:
        docs.append({
            "name": "Udyam MSME Registration Certificate",
            "category": "Business Registration",
            "file_type": "PDF",
            "file_size": "720 KB",
            "status": "Verified",
            "required": True,
            "upload_date": "15 Mar 2024",
            "expiry_date": "Perpetual",
            "renewal_cycle": "Perpetual"
        })

    # Enrich documents with evidence coverage status, mapped statutory requirement, and extracted intelligence
    return enrich_documents_with_evidence_mapping(docs, approvals, profile)


def enrich_documents_with_evidence_mapping(
    docs: List[Dict[str, Any]],
    approvals: List[Dict[str, Any]],
    profile: Any
) -> List[Dict[str, Any]]:
    """
    Enriches each statutory document requirement with evidence coverage status,
    cross-checked approval linkage, and deterministic extracted intelligence.
    """
    approval_map = {}
    for a in approvals:
        name = a.get("name", "")
        approval_map[name] = a
        # Also map by category
        approval_map[a.get("category", "")] = a

    enriched = []
    for idx, doc in enumerate(docs, 1):
        doc_copy = dict(doc)
        doc_name = doc_copy.get("name", "")
        doc_cat = doc_copy.get("category", "")
        status = doc_copy.get("status", "Pending")

        # Determine mapped approval
        mapped_name = None
        mapped_authority = "Statutory Regulatory Department"
        for a in approvals:
            a_name = a.get("name", "")
            if any(k in doc_name.lower() for k in a_name.lower().split()[:2]):
                mapped_name = a_name
                mapped_authority = a.get("authority", mapped_authority)
                break
        if not mapped_name and approvals:
            mapped_name = approvals[idx % len(approvals)].get("name")
            mapped_authority = approvals[idx % len(approvals)].get("authority", mapped_authority)

        # Evidence status: Supported, Missing, Needs Review, Expired
        if status == "Verified":
            evidence_status = "Supported"
            confidence = 0.98
        elif "Audit" in doc_name or "Report" in doc_name or "Notice" in doc_name:
            evidence_status = "Needs Review"
            confidence = 0.85
        else:
            evidence_status = "Missing"
            confidence = 0.65

        # Extracted intelligence from Document Intelligence OCR
        cert_seed = abs(hash(doc_name)) % 90000 + 10000
        extracted_intelligence = {
            "document_type": doc_cat,
            "certificate_number": f"STAT-IND-{cert_seed}",
            "issue_date": doc_copy.get("upload_date", "10 Jan 2024"),
            "expiry_date": doc_copy.get("expiry_date", "Perpetual"),
            "issuing_authority": mapped_authority,
            "confidence_score": confidence,
            "verification_state": "VERIFIED SOURCE"
        }

        doc_copy["mapped_approval"] = mapped_name or "General Statutory Compliance"
        doc_copy["evidence_status"] = evidence_status
        doc_copy["extracted_intelligence"] = extracted_intelligence
        doc_copy["verification_state"] = "VERIFIED SOURCE"

        enriched.append(doc_copy)

    return enriched

File: src/engine/documentRules.js
/**
 * POWER HOUSE Business Analysis Engine - Document Requirement Rules
 */

export function runDocumentRequirementRules(profile, matchedApprovals) {
  const approvalNames = new Set(matchedApprovals.map((a) => a.name));
  const docs = [];

  // 1. Factory License Documents
  if (approvalNames.has('Factory License')) {
    docs.push({
      id: 'gen-doc-factory-cert',
      name: 'Factory Licence Certificate',
      category: 'Statutory License',
      relatedApproval: 'Factory License',
      required: true,
      status: 'Verified',
      fileType: 'PDF',
      fileSize: '2.4 MB',
      uploadDate: '14 Apr 2024',
      expiryDate: '31 Dec 2026',
      description: 'Endorsed Factory License issued by Directorate of Industrial Safety.',
    });
    docs.push({
      id: 'gen-doc-factory-layout',
      name: 'Approved Factory Machinery Layout Plan',
      category: 'Statutory License',
      relatedApproval: 'Factory License',
      required: true,
      status: 'Verified',
      fileType: 'PDF',
      fileSize: '5.1 MB',
      uploadDate: '10 Feb 2024',
      expiryDate: 'Perpetual',
      description: 'Architectural blueprint with machine foundations and clearance corridors.',
    });
  }

  // 2. GST Documents
  if (approvalNames.has('GST Registration & Compliance')) {
    docs.push({
      id: 'gen-doc-gst-cert',
      name: 'GST Registration Certificate (REG-06)',
      category: 'Taxation',
      relatedApproval: 'GST Registration & Compliance',
      required: true,
      status: 'Verified',
      fileType: 'PDF',
      fileSize: '1.1 MB',
      uploadDate: '10 Jan 2024',
      expiryDate: 'Perpetual',
      description: 'Principal place of business registration issued by Central Board of Indirect Taxes.',
    });
  }

  // 3. Pollution Documents
  if (approvalNames.has('Pollution Control NOC (CTO)')) {
    docs.push({
      id: 'gen-doc-pollution-app',
      name: 'Consent to Operate (CTO) Renewal Application',
      category: 'Environmental',
      relatedApproval: 'Pollution Control NOC (CTO)',
      required: true,
      status: 'Pending',
      fileType: 'PDF',
      fileSize: '4.8 MB',
      uploadDate: '02 May 2024',
      expiryDate: 'Under Review',
      description: 'Comprehensive effluent treatment plant (ETP) capacity and air emissions log.',
    });
    docs.push({
      id: 'gen-doc-eia',
      name: 'Environmental Impact Assessment & Hazard Plan',
      category: 'Environmental',
      relatedApproval: 'Pollution Control NOC (CTO)',
      required: true,
      status: 'Required',
      fileType: 'PDF',
      fileSize: '—',
      uploadDate: 'Not Uploaded',
      expiryDate: 'Required for Audit',
      description: 'Detailed analysis of solid and liquid waste generation protocols.',
    });
  }

  // 4. Fire Safety Documents
  if (approvalNames.has('Fire Safety Certificate')) {
    docs.push({
      id: 'gen-doc-fire-report',
      name: 'Fire Safety Hydrant & Sprinkler Audit Report',
      category: 'Safety',
      relatedApproval: 'Fire Safety Certificate',
      required: true,
      status: 'Pending',
      fileType: 'PDF',
      fileSize: '3.2 MB',
      uploadDate: '15 May 2024',
      expiryDate: '15 May 2025',
      description: 'Inspection log of hydraulic water pressure, fire hoses, and alarm sounders.',
    });
  }

  // 5. Labour Documents
  if (approvalNames.has('Employees ESI Registration') || approvalNames.has('Labour Establishment Registration')) {
    docs.push({
      id: 'gen-doc-esi-roster',
      name: 'Employee Master Roster & Form 1 Register',
      category: 'Labor & HR',
      relatedApproval: 'Labour Establishment Registration',
      required: true,
      status: 'Verified',
      fileType: 'XLSX',
      fileSize: '850 KB',
      uploadDate: '12 May 2024',
      expiryDate: '31 Mar 2025',
      description: 'Complete employee wage register, attendance muster, and insurance sub-codes.',
    });
  }

  // 6. MSME Documents
  if (approvalNames.has('MSME Udyam Registration')) {
    docs.push({
      id: 'gen-doc-udyam-cert',
      name: 'Udyam MSME Registration Certificate',
      category: 'Business Registration',
      relatedApproval: 'MSME Udyam Registration',
      required: true,
      status: 'Verified',
      fileType: 'PDF',
      fileSize: '720 KB',
      uploadDate: '15 Mar 2024',
      expiryDate: 'Perpetual',
      description: 'Official Udyam classification as Medium Manufacturing Enterprise.',
    });
  }

  return docs;
}
